src/uct/tcp/tcp_ep.c

```diff
@@ -1522,6 +1580,16 @@ ucs_status_t uct_tcp_ep_flush(uct_ep_h tl_ep, unsigned flags,
     uct_tcp_ep_t *ep = ucs_derived_of(tl_ep, uct_tcp_ep_t);
     uct_tcp_ep_put_completion_t *put_comp;
 
+    if (ucs_unlikely(flags & UCT_FLUSH_FLAG_CANCEL)) {
+        /* TCP is able to cancel only pending operations, posted TX operations
+         * couldn't be canceled, since some data was already sent to the peer
+         * and the peer is waiting for the remaining part of the data */
+        uct_ep_pending_purge(tl_ep,
+                             (uct_pending_purge_callback_t)ucs_empty_function,
+                             0);
```

Comment:
@dmitrygx @evgeny-leksikov in case of UCT_FLUSH_FLAG_CANCEL, do we need tcp to call uct_set_ep_failed()? 
like rc does here - https://github.com/openucx/ucx/blob/master/src/uct/ib/rc/accel/rc_mlx5_ep.c#L546 which eventually calls uct_ep_disconnect for the rdmacm endpoint (in case of closing the ucp ep with mode=force).

Reply:
> in case of UCT_FLUSH_FLAG_CANCEL, do we need tcp to call uct_set_ep_failed()?

@alinask yes, it seems that we need to call it in TCP
@evgeny-leksikov is it correct?

Reply:
actually not, as discussed offline we have to purge and complete all pending/outstanding ops with CANCEL status

Reply:
> actually not, as discussed offline we have to purge and complete all pending/outstanding ops with CANCEL status

@evgeny-leksikov yes, the previous comment was written before the offline discussion

Reply:
uct_ep_pending_purge(tl_ep, NULL, 0);

Reply:
https://github.com/openucx/ucx/pull/5217#discussion_r432922713
`uct_pending_queue_purge()` doesn't support it
we decided to pass empty function instead

Reply:
but it's done in uct_rc_mlx5_ep_flush for example

Reply:
yes, it was my point
RC and others rely on `ucs_arbiter_group_purge` that allows `NULL` callback

Reply:
ok